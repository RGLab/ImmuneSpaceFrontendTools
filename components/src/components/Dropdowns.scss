// Top level dropdown
.df-outer-dropdown {
  height: auto;
  width: min-content;

  & >div>button i {
    padding-left: 5px;
  }

  & > div {
    & > .dropdown-menu {
      display: none;
      width: max-content;
    }
  }

  & > div.open {
    & > .dropdown-menu {
      display: block;
      width: max-content;
    }
  }
}

.df-dropdown-option {
  height: auto;

  & > a {
    padding: 0;

    & > button {
      background: none;
      border: none;
      float: inherit;
      background: none;
      padding: 3px 28px;

      & > i {
        padding-left: 10px;
      }
    }
  }
}



// Nested dropdown (like in manageGroups)
.df-sub-dropdown {
  width: 100%;
  font-size: inherit;
  font-style: inherit;
  font-weight: inherit;
  background: none;

  & > div {
    width: 100%;
    padding: 0;
    color: #494949;

    &.open {
      & > .dropdown-menu {
        display: contents;
      }
    }

    & > button { 
      width: 100%;
      text-align: left;
      background: none;
      padding: 3px 33px;
      border: none;
      font-weight: normal;

      &:focus {
        outline: none;
      }

      &:hover {
        color: #f7f7f7;
        text-decoration: none;
        background-color: #3495d2; // This is the color on prod. It will look weird on test/local where colors are different
      }

    }
  }
}
