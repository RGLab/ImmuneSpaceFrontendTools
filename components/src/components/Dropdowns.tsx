import * as React from 'react'
import "./Dropdowns.scss"

interface ButtonData {
    label: string;
    icon?: JSX.Element;
    action?: () => void;
    disabled?: boolean;
    href?: string;
    buttonData?: ButtonData[]
}

interface DropdownButtonProps {
title: string
buttonData: ButtonData[]
disabled?: boolean
id?: string
}

interface DropdownContentProps {
buttonData: ButtonData[]
open?: boolean
}

interface OuterDropdownButtonProps {
disabled?: boolean;
title: string;
id?: string;
}



// This essentially just creates a bootstrap button dropdown
export const SimpleDropdown: React.FC<DropdownButtonProps> = ({ title, buttonData, disabled, id }) => {
    return (
        <div className="dropdown df-outer-dropdown" id={id}>
        <div className={"btn"} role="group" >
            <button className="btn btn-default dropdown-toggle" data-toggle="dropdown" type="button" disabled={disabled}>
                <span>{title}</span>
                <span style={{paddingLeft:"5px"}}><i className="fa fa-caret-down"></i></span>
            </button>
            <ul className="dropdown-menu df-dropdown">
            {buttonData?.map((button) => {
                return (
                    <li key={button.label} className={"df-dropdown-option"}>
                        <a style={{ padding: "0px 5px" }} onClick={button.action} href={button.href}>
                            {button.label}
                        </a>
                    </li>
                )
            })}
            </ul>
        </div>
    </div>
    )
}


// Dropdown with nesting option
export const DropdownButtons: React.FC<DropdownButtonProps> = ({ title, buttonData, disabled, id }) => {

    return (
        <OuterDropdownButton title={title} disabled={disabled} id={id}>
            <DropdownContent buttonData={buttonData} />
        </OuterDropdownButton>
    )
}

export const OuterDropdownButton: React.FC<OuterDropdownButtonProps> = ({children, disabled, title, id}) => {
    const openRef = React.useRef<HTMLDivElement>(null)
    const open = () => {
        const cl = openRef.current.classList
        const willOpen = !cl.contains("open")
        for (let el of document.querySelectorAll(".df-outer-dropdown>.open")) {
            el.classList.remove("open")
        };
        if (willOpen) {
            cl.add("open")
        }
    }

    return (
        <div className="dropdown df-outer-dropdown" id={id}>
            <div className={"btn"} ref={openRef} role="group" >
                <button className="btn btn-default dropdown-toggle" type="button" disabled={disabled} onClick={open}>
                    <span>{title}</span>
                    <span><i className="fa fa-caret-down"></i></span>
                </button>
                {children}
            </div>
        </div>
    )
}

export const InnerDropdownButtons: React.FC<DropdownButtonProps> = ({ title, buttonData }) => {
    const [open, setOpen] = React.useState(false);
    return (
        <div className="dropdown df-sub-dropdown">
            <div className={open ? " open" : ""} role="group" >
                <button type="button" onClick={() => { setOpen(!open)}}>
                    <span>{title}</span>
                    <span style={{paddingLeft:"5px"}}><i className={"fa fa-caret-"+(open ? "down" : "right")}></i></span>
                </button>
                <InnerDropdownContent buttonData={buttonData} open={open} />
            </div>
        </div>
    )
}

const DropdownContent: React.FC<DropdownContentProps> = ({buttonData}) => {
    return <ul className="dropdown-menu df-dropdown">
    {buttonData.map((button) => {
        return (
            <li key={button.label} className={"df-dropdown-option " + (button.disabled ? "disabled" : "")}>
                {(() => {
                    if (button.buttonData) {
                        return (<InnerDropdownButtons title={button.label} buttonData={button.buttonData}/>);
                    } else {
                        return <a style={{ padding: "0px 5px" }} key={button.label} href={button.href}>
                            <button onClick={button.action ?? (() => { })} >
                                {button.label}{button.icon}
                            </button>
                        </a>
                    }
                })()
                }

            </li>
        )
    })}
</ul>
}

const InnerDropdownContent: React.FC<DropdownContentProps> = ({buttonData, open}) => {
    return <ul className="dropdown-menu df-dropdown" style={open ? {display: "contents"} : {}}>
    {buttonData.map((button) => {
        return (
            <li key={button.label} className={"df-dropdown-option " + (button.disabled ? "disabled" : "")}>
                {(() => {
                    if (button.buttonData) {
                        return (<InnerDropdownButtons title={button.label} buttonData={button.buttonData}/>);
                    } else {
                        return <a style={{ padding: "0px 5px", marginLeft: "20px" }} key={button.label} href={button.href}>
                            <button onClick={button.action ?? (() => { })} >
                                {button.label}{button.icon}
                            </button>
                        </a>
                    }
                })()
                }

            </li>
        )
    })}
</ul>
}
