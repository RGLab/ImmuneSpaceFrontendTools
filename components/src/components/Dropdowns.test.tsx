import React from 'react'
import * as renderer from 'react-test-renderer'
import * as dropdowns from './Dropdowns'
import { mount } from "enzyme";

describe("<InnerDropdownButtons />", () => {
    const buttonData = [
        {
            label: "option1"
        },{
            label: "option2",
            buttonData: [{label: "option3"}]
        }
    ]
    test("Render", () => {
        const tree = renderer.create(<dropdowns.InnerDropdownButtons title={"test"} buttonData={buttonData} />).toJSON()
        expect(tree).toMatchSnapshot()
    })
    test("Content", () => {
        const dropdown = mount(<dropdowns.InnerDropdownButtons title={"test"} buttonData={buttonData} />)
        expect(dropdown.find("button").length).toEqual(4)
        expect(dropdown.find("InnerDropdownContent").length).toEqual(2)
        expect(dropdown.find("InnerDropdownContent").first().children().find("InnerDropdownContent").length).toEqual(1)
    })
    test("open", () => {
        const dropdown = mount(<dropdowns.InnerDropdownButtons title={"test"} buttonData={buttonData} />)
        expect(dropdown.find("InnerDropdownContent").first().prop("open")).toBeFalsy()
        dropdown.find("button").first().simulate("click")
        expect(dropdown.find("InnerDropdownContent").first().prop("open")).toBeTruthy()
    })
})

describe("<OuterDropdownButtons />", () => {
    test("Render", () => {
        const tree = renderer.create(
            <dropdowns.OuterDropdownButton title={"Test"}><div id="test-div"/></dropdowns.OuterDropdownButton>
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    test("Children", () => {
        const button = mount(
            <dropdowns.OuterDropdownButton title={"Test"}><div id="test-div"/></dropdowns.OuterDropdownButton>
        )
        expect(button.find(".df-outer-dropdown").length).toEqual(1)
        expect(button.find(".df-outer-dropdown>.btn").children().length).toEqual(2)
        expect(button.find(".df-outer-dropdown>.btn").children().last().html()).toEqual('<div id="test-div"></div>')
    })
})


describe("<DropdownButtons />", () => {
    const clickMock1 = jest.fn()
    const clickMock2 = jest.fn()
    const buttonData = [
        {
            label: "Single Button",
            action: clickMock1
        },
        {
            label: "Nest",
            buttonData: [
                {
                    label: "egg1",
                    href: "#"
                },
                {
                    label: "another nest",
                    buttonData: [
                        {label:"super nested egg", action: clickMock2}
                    ]
                },
                {
                    label: "egg2"
                }
            ]
        }
    ]
    test("Render", () => {
        const tree = renderer.create(<dropdowns.DropdownButtons title={"Save"} buttonData={buttonData} />).toJSON()
        expect(tree).toMatchSnapshot()
    })

    test("Contents", () => {
        const nestedDropdown = mount(<dropdowns.DropdownButtons title={"Dropdown"} buttonData={buttonData}/>)
        expect(nestedDropdown.find("button").length).toEqual(7)
        expect(nestedDropdown.find(".dropdown-menu").length).toEqual(3)
        expect(nestedDropdown.find(".df-sub-dropdown").length).toEqual(2)
        expect(nestedDropdown.find("a").length).toEqual(4)
        expect(nestedDropdown.find('a[href="#"]').length).toEqual(1)
    })

    test("actions", () => {
        const nestedDropdown = mount(<dropdowns.DropdownButtons title={"Dropdown"} buttonData={buttonData}/>)
        
        expect(clickMock1.mock.calls.length).toEqual(0)
        nestedDropdown.findWhere(x => x.text() == "Single Button").find("button").simulate("click")
        expect(clickMock1.mock.calls.length).toEqual(1)

        expect(clickMock2.mock.calls.length).toEqual(0)
        nestedDropdown.findWhere(x => x.text() == "super nested egg").find("button").simulate("click")
        expect(clickMock2.mock.calls.length).toEqual(1)
    })
})

