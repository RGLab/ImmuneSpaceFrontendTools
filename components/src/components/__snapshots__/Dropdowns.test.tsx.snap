// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownButtons /> Render 1`] = `
<div
  className="dropdown df-outer-dropdown"
>
  <div
    className="btn"
    role="group"
  >
    <button
      className="btn btn-default dropdown-toggle"
      onClick={[Function]}
      type="button"
    >
      <span>
        Save
      </span>
      <span>
        <i
          className="fa fa-caret-down"
        />
      </span>
    </button>
    <ul
      className="dropdown-menu df-dropdown"
    >
      <li
        className="df-dropdown-option "
      >
        <a
          style={
            Object {
              "padding": "0px 5px",
            }
          }
        >
          <button
            onClick={[MockFunction]}
          >
            Single Button
          </button>
        </a>
      </li>
      <li
        className="df-dropdown-option "
      >
        <div
          className="dropdown df-sub-dropdown"
        >
          <div
            className=""
            role="group"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              <span>
                Nest
              </span>
              <span
                style={
                  Object {
                    "paddingLeft": "5px",
                  }
                }
              >
                <i
                  className="fa fa-caret-right"
                />
              </span>
            </button>
            <ul
              className="dropdown-menu df-dropdown"
              style={Object {}}
            >
              <li
                className="df-dropdown-option "
              >
                <a
                  href="#"
                  style={
                    Object {
                      "marginLeft": "20px",
                      "padding": "0px 5px",
                    }
                  }
                >
                  <button
                    onClick={[Function]}
                  >
                    egg1
                  </button>
                </a>
              </li>
              <li
                className="df-dropdown-option "
              >
                <div
                  className="dropdown df-sub-dropdown"
                >
                  <div
                    className=""
                    role="group"
                  >
                    <button
                      onClick={[Function]}
                      type="button"
                    >
                      <span>
                        another nest
                      </span>
                      <span
                        style={
                          Object {
                            "paddingLeft": "5px",
                          }
                        }
                      >
                        <i
                          className="fa fa-caret-right"
                        />
                      </span>
                    </button>
                    <ul
                      className="dropdown-menu df-dropdown"
                      style={Object {}}
                    >
                      <li
                        className="df-dropdown-option "
                      >
                        <a
                          style={
                            Object {
                              "marginLeft": "20px",
                              "padding": "0px 5px",
                            }
                          }
                        >
                          <button
                            onClick={[MockFunction]}
                          >
                            super nested egg
                          </button>
                        </a>
                      </li>
                    </ul>
                  </div>
                </div>
              </li>
              <li
                className="df-dropdown-option "
              >
                <a
                  style={
                    Object {
                      "marginLeft": "20px",
                      "padding": "0px 5px",
                    }
                  }
                >
                  <button
                    onClick={[Function]}
                  >
                    egg2
                  </button>
                </a>
              </li>
            </ul>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`<InnerDropdownButtons /> Render 1`] = `
<div
  className="dropdown df-sub-dropdown"
>
  <div
    className=""
    role="group"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      <span>
        test
      </span>
      <span
        style={
          Object {
            "paddingLeft": "5px",
          }
        }
      >
        <i
          className="fa fa-caret-right"
        />
      </span>
    </button>
    <ul
      className="dropdown-menu df-dropdown"
      style={Object {}}
    >
      <li
        className="df-dropdown-option "
      >
        <a
          style={
            Object {
              "marginLeft": "20px",
              "padding": "0px 5px",
            }
          }
        >
          <button
            onClick={[Function]}
          >
            option1
          </button>
        </a>
      </li>
      <li
        className="df-dropdown-option "
      >
        <div
          className="dropdown df-sub-dropdown"
        >
          <div
            className=""
            role="group"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              <span>
                option2
              </span>
              <span
                style={
                  Object {
                    "paddingLeft": "5px",
                  }
                }
              >
                <i
                  className="fa fa-caret-right"
                />
              </span>
            </button>
            <ul
              className="dropdown-menu df-dropdown"
              style={Object {}}
            >
              <li
                className="df-dropdown-option "
              >
                <a
                  style={
                    Object {
                      "marginLeft": "20px",
                      "padding": "0px 5px",
                    }
                  }
                >
                  <button
                    onClick={[Function]}
                  >
                    option3
                  </button>
                </a>
              </li>
            </ul>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`<OuterDropdownButtons /> Render 1`] = `
<div
  className="dropdown df-outer-dropdown"
>
  <div
    className="btn"
    role="group"
  >
    <button
      className="btn btn-default dropdown-toggle"
      onClick={[Function]}
      type="button"
    >
      <span>
        Test
      </span>
      <span>
        <i
          className="fa fa-caret-down"
        />
      </span>
    </button>
    <div
      id="test-div"
    />
  </div>
</div>
`;
